Add ComplexHeatmap to RNAseq

jashapiro opened the pull request:
This PR adds a notebook about generating a ComplexHeatmap to the RNAseq module.

This notebook uses a subset of the OpenPBTA data, and also includes the steps of transforming the raw count data with `vst()` to make them suitable for clustering and display in a heatmap.

Questions that need to be addressed (some in progress as of this draft, so this is partly a progress note to myself):
- Is the explanation of loading data from a table rather than with `tximeta` sufficient?
- How much explanation of VST is required?
- How much clustering should go in this notebook vs. potentially adding one more? 
  - I feel like consensus clustering is perhaps too much, but some dimension reduction might be worthwhile? I'd love to hear thoughts from others.

jaclyn-taroni commented on RNA-seq/06-openpbta_heatmap.Rmd:
```diff
@@ -0,0 +1,430 @@
+
+The `DESeq2` package has functions for doing these calculations, so we will convert our count matrix to a `DESeq2` dataset and perform the calculations there.
+
+Since we are starting from a matrix, not a `SummarizedExperimet` as we did previously, we will need to provide the sample information ourselves.
```
```suggestion
Since we are starting from a matrix, not a `SummarizedExperiment` as we did previously, we will need to provide the sample information ourselves.
```

jaclyn-taroni commented on RNA-seq/06-openpbta_heatmap.Rmd:
```diff
@@ -0,0 +1,430 @@
+To improve this, we will transform the count data to create an expression measure that is better suited for these analyses. 
+The core transformation will map the expression to a log2 scale, while accounting for some of the expected variation among samples and genes.
+
+Since different samples are usually sequenced to different depths, we want to transform our RNA-seq count data to make different samples more directly comparable. 
+We also want to deal with the fact that genes with low counts are also likely to have higher variance (on the log2 scale), as that could bias our clustering.
+To handle both of these considerations, we can calculate a Variance Stabilizing Transformation of the count data, and work with that transformed data for our analysis.
```
This is not the first time they will see this transformation, should this explanation be moved here instead: https://github.com/AlexsLemonade/training-modules/blob/b989631516d393f1707d321107025ec25afc6ee4/RNA-seq/03-gastric_cancer_exploratory.Rmd#L89

We also cover VST in the first RNA-seq exercise notebook. The point there is to visualize what the transformation does to the underlying values! Starts here: https://github.com/AlexsLemonade/training-modules/blob/b989631516d393f1707d321107025ec25afc6ee4/RNA-seq/exercise_01-exploratory_data_analysis.Rmd#L126

jaclyn-taroni replied:
Converting to a `DESeqDataSet` from a matrix here makes sense to me! Because I know that you have another branch with PCA for this dataset, I will go ahead and suggest that should you have put a more detailed description of PCA in that notebook, you can move it to `03` also? I do still think there may be value in comparing and contrasting hierarchical clustering with PCA, time permitting.

jaclyn-taroni commented on RNA-seq/06-openpbta_heatmap.Rmd:
```diff
@@ -0,0 +1,430 @@
+You probably noticed that we a couple options in there when we made the heatmap about "clustering".
+These options refer to the hierarchical clustering method that is used to determine the order of rows and columns in the heatmap, bringing samples and genes that show similar patterns together.
+
+We will explore 
```
In this essay, I will

jaclyn-taroni commented on RNA-seq/06-openpbta_heatmap.Rmd:
````diff
@@ -0,0 +1,430 @@
+dev.off()
+```
+
+### Notes on clustering
````
Yep, I think this might be too long as is. BUT! We should also move your nice explanation of VST out of this notebook in my opinion. If we do that, do you still want do the transformation during the instruction or can we cooking show magic it 🍳 ?

jaclyn-taroni commented on RNA-seq/06-openpbta_heatmap.Rmd:
````diff
@@ -0,0 +1,430 @@
+)
+```
+
+### Values for display
````
This is another section we can chop if we're concerned about length?

jaclyn-taroni commented on RNA-seq/06-openpbta_heatmap.Rmd:
````diff
@@ -0,0 +1,430 @@
+It's calculated for each value in a row (gene) by subtracting the gene's mean and dividing by the gene's standard deviation.
+
+```{r}
+zscores_mat <- 
````
I guess before I say that I should ask: What does the heatmap look like without scaling?

jashapiro replied:
Eh, it isn't great: 
![image](https://user-images.githubusercontent.com/20483/110190958-08ea5800-7df4-11eb-91d7-dd415f9ed048.png)

I bet I can tighten up the z score explanation though.

jaclyn-taroni reviewed (commented):
High-level comments! I think we can move your (more substantial) explanation about VST to the gastric cancer EDA notebook and that may change your mind about performing the transformation during instruction = shorter length.

jashapiro commented on RNA-seq/06-openpbta_heatmap.Rmd:
```diff
@@ -0,0 +1,430 @@
+
+### Convert and round
+
+The count data we have is mostly integers, but because of the estimation procedure that RSEM uses, some is not. 
```
I might rewrite this to be less controversial.

jashapiro reviewed (commented):
Thanks for the review! Helpful to be reminded of some bigger picture things...

jashapiro commented:
I ended up cutting out the medulloblastoma section in favor of a short PCA comparison and some expanded notes. I moved the VST discussion to the `03` exploratory notebook, and added a bit there about PCA as well. 

I think I feel pretty good about the level of content at this point... curious again what others think.

(I also found a version of the blog post @jaclyn-taroni had suggested with less yellow. Oh and https://github.com/AlexsLemonade/training-modules/pull/417/commits/ff4fcfc1f5a70df1f8dfdbda21d5ad0518fc305b)

jaclyn-taroni commented on RNA-seq/03-gastric_cancer_exploratory.Rmd:
```diff
@@ -88,27 +85,33 @@ ddset <- DESeqDataSet(gene_summarized,
+The expression of many of those genes are correlated, so we can often get a better, simpler picture of the data by combining the information from those correlated genes.
+
+PCA rotates and transforms this space so that each axis is now a combination of multiple correlated genes, ordered so the first axes capture the most variation from the data. 
+These new axes are the "principal components".
```
```suggestion
These new axes are the "principal components."
```

jaclyn-taroni commented on RNA-seq/06-openpbta_heatmap.Rmd:
````diff
@@ -0,0 +1,385 @@
+```{r color_palettes}
+# The Okabe Ito palette is recommended for those with color vision deficiencies  
+histology_colors <- palette.colors(palette = "Okabe-Ito")[2:5]
+# `palette.colors()` returns a named vector, which can cause trouble
````
Only sometimes 😈

jaclyn-taroni commented on RNA-seq/06-openpbta_heatmap.Rmd:
````diff
@@ -0,0 +1,385 @@
+  (high_var_mat - rowMeans(high_var_mat)) / matrixStats::rowSds(high_var_mat)
+```
+
+Since all genes have the same z-score variance (by definition!), if you're filtering to high variance values, it's important to do that _prior_ to standardizing.
````
👍

jaclyn-taroni commented on RNA-seq/06-openpbta_heatmap.Rmd:
```diff
@@ -0,0 +1,385 @@
+We defined the distance between samples as 1 minus the correlation between them, and then averaged samples when they were merged into a group.
+The method we're using is also known as [UPGMA (unweighted pair group method with arithmetic mean)](https://en.wikipedia.org/wiki/UPGMA).
+
+It is important to note that the linear arrangement of samples in the the tree is somewhat arbitrary, so we have to be careful not to overintepret it.
```
```suggestion
It is important to note that the linear arrangement of samples in the the tree is somewhat arbitrary, so we have to be careful not to overinterpret it.
```

jaclyn-taroni commented on RNA-seq/06-openpbta_heatmap.Rmd:
```diff
@@ -0,0 +1,385 @@
+We defined the distance between samples as 1 minus the correlation between them, and then averaged samples when they were merged into a group.
+The method we're using is also known as [UPGMA (unweighted pair group method with arithmetic mean)](https://en.wikipedia.org/wiki/UPGMA).
+
+It is important to note that the linear arrangement of samples in the the tree is somewhat arbitrary, so we have to be careful not to overintepret it.
+We can rotate around any branch of the tree in our visualization with no change to the tree topology itself.
+In fact, the samples at opposite ends of the heatmap could be right next to one another with the right set of rotations!
```
Do you want to talk about branch lengths here? Caveat: I don't know much about dendrogram defaults in `ComplexHeatmap`.

jashapiro replied:
No, I do not want to talk about branch lengths.

jaclyn-taroni commented on RNA-seq/06-openpbta_heatmap.Rmd:
```diff
@@ -0,0 +1,385 @@
+The kind of clustering we performed was [hierarchical clustering](https://en.wikipedia.org/wiki/Hierarchical_clustering).
+This is an _agglomerative_ method of clustering - each sample starts in its own cluster and then at each step of the algorithm the two most similar clusters are joined until there's only a single cluster left.
+
+We defined the distance between samples as 1 minus the correlation between them, and then averaged samples when they were merged into a group.
```
I'm not sure _why_ but "We defined..." strikes me as odd here since it's happening under the hood. This opinion is very weakly held.

jashapiro replied:
We defined it with the `clustering_distance_rows` argument, is what I was trying to get across.

jaclyn-taroni commented on RNA-seq/06-openpbta_heatmap.Rmd:
````diff
@@ -0,0 +1,385 @@
+        name = "z-score")
+```
+
+We have to create the heatmap _twice_ -- once for display in the notebook and once to save to a PNG file.
````
Is this still true? I can never keep up.

jashapiro replied:
I tried to find a way not to do this, but didn't. 

I also had a twitter conversation with Jokergoo about a regression in the current version where you can't customize rasterizing. He implemented "better" rasterization without thinking it through for all cases, so now you get good screen resolution rasterization, but can't push resolution higher for output except by turning off rasterization... The net result is that I am actually doing something a bit different for the output block (`use_raster = FALSE`).

jaclyn-taroni commented on RNA-seq/06-openpbta_heatmap.Rmd:
````diff
@@ -0,0 +1,385 @@
+
+```{r expr-pca }
+# Use plotPCA, but return the data for custom plotting
+pca_df <- plotPCA(vst_data, 
+        ntop = 5000, # use the top 5000 genes by variance
+        intgroup = "short_histology",
+        returnData = TRUE) 
````
Something going on with the spacing here.

jaclyn-taroni commented on RNA-seq/06-openpbta_heatmap.Rmd:
```diff
@@ -0,0 +1,385 @@
+  )
+)
+
+# Customize the legend titles
+legend_titles =  list(
+    short_histology = list(title = "Histology"),
+    composition = list(title = "Composition"),
+    tumor_descriptor = list(title = "Tumor descriptor")
+    )
```
```suggestion
# Customize the legend titles
legend_titles =  list(
      short_histology = list(title = "Histology"),
      composition = list(title = "Composition"),
      tumor_descriptor = list(title = "Tumor descriptor")
    )
```
Or is this a `styler` weird choice? It has that vibe.

jashapiro replied:
It was some kind of a pasted spacing bug, but more importantly, I went to another method to do this, so this section is irrelevant.

jaclyn-taroni reviewed (commented):
Much tighter! I, too, like this level of content for this final notebook. Most of my comments are very minor. I did not run this myself and look at the plots; it did not seem necessary for a draft.

Side note: this diff looks a bit 🍌 and I'm not sure why - it looks like you merged `master` into both branches involved to me?

jashapiro commented:
> Side note: this diff looks a bit 🍌 and I'm not sure why - it looks like you merged `master` into both branches involved to me?

I thought I did... can't quite figure out what is going on there. I think I will see what happens after #415 goes in, then deal with whatever lingering extra changes there are.

jashapiro commented:
> > Side note: this diff looks a bit 🍌 and I'm not sure why - it looks like you merged `master` into both branches involved to me?
> 
> I thought I did... can't quite figure out what is going on there. I think I will see what happens after #415 goes in, then deal with whatever lingering extra changes there are.

Seems to have resolved itself. Good work everyone!

jashapiro commented on RNA-seq/06-openpbta_heatmap.Rmd:
````diff
@@ -0,0 +1,379 @@
+        # This will be used as the label for the color bar
+        # of the cells of the heatmap itself
+        name = "z-score")
+```
````
I was debating about this myself, but I think I decided I just don't want to get into it! I kind of hate the message, and I really don't want to get into how to fix it, but I might be convinced to reconsider. (I also hate that the two messages give different instructions for how to hide them.)

jashapiro replied:
I think I will add a note to expect some warnings.

jashapiro commented on RNA-seq/06-openpbta_heatmap.Rmd:
````diff
@@ -0,0 +1,379 @@
+        # of the cells of the heatmap itself
+        name = "z-score")
+```
+
````
I think the main things I wanted to cover were in the "Notes on clustering" section, which is what I was thinking about as the main way to address the general interpretation questions. Was there something else specific you were thinking we should add or clarify? 

Looking over that, I don't think I ever say explicitly that the samples and genes are arranged by the clustering algorithm, so I should probably say that up top. Other suggestions?

jashapiro replied:
I added some discussion of ordering (and warnings). Let me know what you think!
